apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -440,11 +496,63 @@ const DanceAiModal: React.FunctionComponent = () => {
       return <span key={index}>{part}</span>;
     });
 
-  const currentGeneratedEffect = getCurrentGeneratedEffect();
+  const currentGeneratedEffect = getGeneratedEffect(generatingProgress.step);
 
   const lastInputItem =
     currentInputSlot > 0 ? getItem(inputs[currentInputSlot - 1]) : undefined;
 
+  // Visualization preview size, in pixels.
+  const previewSize = 280;
+  const previewSizeSmall = 90;
+
+  const labels = getLabels();
+
+  // Calculates the minimum individual score
+  // (ie, a SINGLE emoji association with a foreground/background palette combination),
+  // and a maximum total score
+  // (ie, the sum of ALL selected emoji's associations with a foreground/background palette combination).
+  // Used to normalize and scale the data for easier differentiation between results by the user.
+  const calculateMinMax = () => {
```

Comment:
Does this run on every step/substep? And if so, is this tracking min/max values across all generated effects? Trying to understand the relationship between what's calculated here and the `<Score>` that's displayed with the currently showing effect.

Reply:
I think so.  I'm not so concerned about the cost of executing this when we re-render, but it does seem a valid simplification and optimization to regenerate this only when needed.  @bencodeorg to consider.

Reply:
Yes re: tracking min/max values across all generated effects. This really only needs to run in two places (once after we select effects in `startAi`, and once anytime we reopen the modal and regenerate the random "bad" effects you were looking at above).

I think (correct me if I'm wrong), the right way to do this would be to use useEffect and have it only regenerate when the good/bad effects change. When I tried to do this, I ran into a bit of a tangled web of dependencies between different functions, linter suggestions to move functions into useCallback, etc. It seemed not so expensive (and would have produced a lot of merge conflicts for Brendan to sort out if I did refactor), so I implemented this way. Definitely think this could be revisited!

Reply:
Gotcha, yeah my confusion was more around - when we're running through each of the bad effects, where are we computing the values to show for that specific score? If this calculation runs on all generated effects, how are we getting the individual numbers to show the bar for each effect?

Reply:
mmm yeah good q -- I think this is where we're computing values first:

https://github.com/code-dot-org/code-dot-org/blob/cae4e03c447ce0aee38c8f00cab24aae8194fa60/apps/src/dance/ai/DanceAiModal.tsx#L768-L779

Then we calculate a "net score" using the min/max to actually graph them:

https://github.com/code-dot-org/code-dot-org/blob/cae4e03c447ce0aee38c8f00cab24aae8194fa60/apps/src/dance/ai/DanceAiModal.tsx#L953-L997

There's gotta be a cleaner way to do this, but alas. In any case, I think we need two passes -- one to loop through the data we want to visualize to establish the min and max so we can rescale the visualization, and another pass to actually do the visualization.

Reply:
Oh I see, that makes sense! Thanks for clarifying

Reply:
We generate the effects by calling `chooseEffects` in two places:  in normal flows [here](https://github.com/code-dot-org/code-dot-org/blob/b1b8a5b246b21706a56e6fdf6669de242ec5d188/apps/src/dance/ai/DanceAiModal.tsx#L325-L327); and when reopening the modal with an existing "good" effect [here](https://github.com/code-dot-org/code-dot-org/blob/b1b8a5b246b21706a56e6fdf6669de242ec5d188/apps/src/dance/ai/DanceAiModal.tsx#L164).  

Once we know a chosen effect and the three emoji, we can call `getGeneratedEffectScores` to get the scores that are used in rendering the bars.  

We'd need to be careful about making sure we're using the right values (and not, say, relying on a `useState` value that is pending to be set) but we should be able to cache the scores at those two places.

Reply:
This turned out to be a prescient thread, with https://github.com/code-dot-org/code-dot-org/pull/54850 being a proposed attempt at this.